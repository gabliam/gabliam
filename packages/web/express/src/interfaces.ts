import { DefaultMethods } from '@gabliam/web-core';
import { express } from './express';

/**
 * Represent a method that create an express router
 */
export type RouterCreator = () => express.Router;

export type ExpressMethods = Extract<keyof express.Router, DefaultMethods>;
