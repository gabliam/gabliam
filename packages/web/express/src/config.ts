import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import bodyParser from 'body-parser';
import { EXPRESS_PLUGIN_CONFIG } from './constants';

const options = Joi.object().keys({
  inflate: Joi.boolean(),
  limit: [Joi.number(), Joi.string()],
  type: [Joi.string(), Joi.array().items(Joi.string())],
});

const optionsJson = options.keys({
  strict: Joi.boolean(),
});

const optionsText = options.keys({
  defaultCharset: Joi.string(),
});

const optionsUrlencoded = options
  .keys({
    extended: Joi.boolean(),
    parameterLimit: Joi.number(),
  })
  .default({ extended: false });

const configJoi = Joi.object()
  .keys({
    json: Joi.boolean().default(true),
    optionsJson,
    urlencoded: Joi.boolean().default(true),
    optionsUrlencoded,
    text: Joi.boolean().default(false),
    optionsText,
  })
  .default({
    json: true,
    urlencoded: true,
    optionsUrlencoded: { extended: false },
    text: false,
  });

export interface BodyParserConfig {
  json: boolean;
  optionsJson?: bodyParser.OptionsJson;
  urlencoded: boolean;
  optionsUrlencoded?: bodyParser.OptionsUrlencoded;
  text: boolean;
  optionsText?: bodyParser.OptionsText;
}

export interface ExpressConfig {
  bodyParser: BodyParserConfig;
}

@PluginConfig()
export class ExpressPluginConfig implements ExpressConfig {
  @Value('application.web.bodyParser', configJoi)
  bodyParser: BodyParserConfig;

  @Bean(EXPRESS_PLUGIN_CONFIG)
  restConfig(): ExpressConfig {
    return {
      bodyParser: this.bodyParser,
    };
  }
}
