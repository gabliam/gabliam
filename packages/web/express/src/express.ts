import express from 'express';
import { GabContext } from '@gabliam/web-core';

export { express };

export interface CustomRequest extends express.Request {
  context: GabContext;
}
