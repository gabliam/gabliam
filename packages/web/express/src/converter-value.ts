import {
  ExecutionContext,
  GabContext,
  ResponseEntity,
  convertValueFn,
} from '@gabliam/web-core';
import { express } from './express';

export const converterValue: convertValueFn = (
  ctx: GabContext,
  execCtx: ExecutionContext,
  result: any,
) => {
  // response handler if the result is a ResponseEntity
  function convertEntity(value: ResponseEntity) {
    if (value.hasHeader()) {
      ctx.response.set(<any>value.headers);
    }
    ctx.status = value.status;
    ctx.body = value.body;
  }

  if (!ctx.headersSent) {
    if (result !== undefined) {
      if (result instanceof ResponseEntity) {
        convertEntity(result);
      } else if (typeof result === 'string' || typeof result === 'object') {
        ctx.body = result;
      } else {
        ctx.body = `${result}`;
      }
    }
  }
};

export const send = (ctx: GabContext, res: express.Response, json: boolean) => {
  const { status, message, body, type } = ctx;
  /* istanbul ignore next */
  if (type) {
    res.type(type);
  }
  /* istanbul ignore next */
  if (message) {
    res.statusMessage = message;
  }

  if (status) {
    res.status(status);
  }
  if (json) {
    res.json(body);
  } else {
    res.send(body);
  }
};
