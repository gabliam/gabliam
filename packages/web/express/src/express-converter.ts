import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  BadInterceptorError,
  createInterceptorResolver,
  getContext,
  getExtractArgs,
  isInterceptor,
} from '@gabliam/web-core';
import { express } from './express';

/**
 * Class for convert interceptor to Express Middleware
 */
@InjectContainer()
@Service()
export class ExpressConverter {
  public interceptorToMiddleware(clazz: any): express.RequestHandler {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new BadInterceptorError(instance);
    }

    const extractArgs = getExtractArgs(container, instance, 'intercept', true);

    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction,
    ) => {
      const args = await extractArgs(getContext(req), null, next);

      try {
        await toPromise(instance.intercept(...args));
        next();
      } catch (err) {
        next(err);
      }
    };
  }
}
