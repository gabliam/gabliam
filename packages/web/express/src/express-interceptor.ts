import { Service } from '@gabliam/core';
import {
  Context,
  GabContext,
  Interceptor,
  InterceptorConstructor,
  METADATA_KEY,
  Next,
  nextFn,
  UseInterceptors,
} from '@gabliam/web-core';
import { express } from './express';

export type ExpressInterceptorType = 'RequestHandler' | 'intercept';

/**
 * Test if target is an express interceptor
 * @param target any
 */
export const isValidInterceptor = (target: any) => {
  const meta = Reflect.getOwnMetadata(
    METADATA_KEY.specialInterceptor,
    target.constructor || target,
  );

  return meta === undefined || meta === 'express';
};

/**
 * ExpressInterceptor decorator
 * class is an express interceptor.
 * Must return express middleware
 */
const ExpressInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.specialInterceptor, 'express', target);
};

/**
 * Convert a Express middleware to an express interceptor
 */
export const toInterceptor = (
  mid: express.RequestHandler,
): InterceptorConstructor => {
  const clazz: InterceptorConstructor = class implements Interceptor {
    async intercept(context: GabContext, next: nextFn): Promise<any> {
      const req = context.request.originalRequest;
      const res = context.response.originalResponse;
      switch (mid.length) {
        // with callback
        case 3:
          return new Promise((resolve, reject) => {
            (<express.RequestHandler>mid)(req, res, (err: any) => {
              if (err) {
                reject(err);
              } else {
                resolve(next());
              }
            });
          });
        // with error
        case 2:
        default:
          (<any>mid)(req, res);
          await next();
          return Promise.resolve();
      }
    }
  };
  Context()(clazz.prototype, 'intercept', 0);
  Next()(clazz.prototype, 'intercept', 1);
  Service()(clazz);
  ExpressInterceptor()(clazz);

  // metadata
  Reflect.defineMetadata('design:type', Function, clazz.prototype, 'intercept');
  Reflect.defineMetadata(
    'design:paramtypes',
    [Function, Function],
    clazz.prototype,
    'intercept',
  );
  Reflect.defineMetadata(
    'design:returntype',
    Promise,
    clazz.prototype,
    'intercept',
  );
  return clazz;
};

/**
 * Alias for evict to use: UseInterceptors(toInterceptor(expressMid))
 * /!\ Use it for RequestHandler
 */
export const UseExpressInterceptors = (...mids: express.RequestHandler[]) =>
  UseInterceptors(...mids.map((m) => toInterceptor(m)));
