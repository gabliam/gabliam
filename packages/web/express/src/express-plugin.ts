import {
  Container,
  GabliamPlugin,
  Plugin,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import {
  APP,
  cleanPath,
  compose,
  ExecutionContext,
  getContext,
  InterceptorInfo,
  REQUEST_LISTENER_CREATOR,
  RestMetadata,
  SERVER,
  WebConfiguration,
  WebConfigurationContructor,
  WebPluginBase,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
} from '@gabliam/web-core';
import d from 'debug';
import http from 'http';
import { CUSTOM_ROUTER_CREATOR } from './constants';
import { converterValue, send } from './converter-value';
import { express } from './express';
import { isValidInterceptor } from './express-interceptor';
import { ExpressMethods, RouterCreator } from './interfaces';
import { addContextMiddleware, addMiddlewares } from './middleware';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('ExpressPlugin')
@Scan()
export class ExpressPlugin extends WebPluginBase<express.Application>
  implements GabliamPlugin {
  constructor(
    config?: Partial<WebConfigurationContructor<express.Application>>
  ) {
    super(config);
  }

  bindApp(
    container: Container,
    registry: Registry,
    webConfiguration: WebConfiguration
  ): void {
    container.bind(APP).toConstantValue(express());
    container.bind(REQUEST_LISTENER_CREATOR).toConstantValue(() => {
      const app = container.get<express.Application>(APP);
      return app;
    });
    webConfiguration.addwebConfig({
      instance: addMiddlewares,
      order: -2,
    });

    webConfiguration.addwebConfig({
      instance: addContextMiddleware,
      order: -1,
    });
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    app.set('port', restConfig.port);
    await super.start(container, registry);
  }

  async buildControllers(
    restMetadata: RestMetadata<ExpressMethods>,
    container: Container
  ) {
    const app = container.get<express.Application>(APP);

    // get the router creator
    let routerCreator: RouterCreator = () => express.Router();
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    for (const [
      controllerId,
      { methods, controllerPath },
    ] of restMetadata.controllerInfo) {
      const controller = container.get<object>(controllerId);
      const router = routerCreator();
      const routerPath = cleanPath(`${restMetadata.rootPath}${controllerPath}`);

      debug(`New route : "${routerPath}"`);

      for (const methodInfo of methods) {
        const execCtx = new ExecutionContext(controller, methodInfo);

        const interceptors = methodInfo.interceptors.filter(i =>
          isValidInterceptor(i)
        );

        // create handler
        const handler = this.handlerFactory(execCtx, interceptors);

        // register handler in router
        router[methodInfo.method](methodInfo.methodPath, handler);
      }
      app.use(routerPath, router);
    }
  }

  private handlerFactory(
    execCtx: ExecutionContext,
    interceptors: InterceptorInfo[]
  ): express.RequestHandler {
    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const composeInterceptor = compose(
        interceptors,
        converterValue
      );
      const ctx = getContext(req);
      const methodInfo = execCtx.getMethodInfo();
      const controller = execCtx.getClass();

      let expressNextCall = false;
      const expressNext = () => {
        expressNextCall = true;
        next();
      };

      const callNext = async () => {
        const args = await toPromise(
          methodInfo.extractArgs(ctx, execCtx, expressNext)
        );
        return await toPromise(controller[methodInfo.methodName](...args));
      };

      try {
        await composeInterceptor(ctx, execCtx, callNext);
        if (!expressNextCall) {
          send(ctx, res, methodInfo.json);
        }
      } catch (err) {
        next(err);
      }
    };
  }
}
