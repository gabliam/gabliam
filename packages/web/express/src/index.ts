import { ExpressPlugin } from './express-plugin';

export * from 'http-status-codes';
export * from './express';
export * from './express-converter';
export * from './express-interceptor';
export * from './interfaces';
export * from './request';
export * from './response';
export { ExpressPlugin as default };
