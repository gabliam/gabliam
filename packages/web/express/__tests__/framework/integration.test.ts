import { Bean, Config } from '@gabliam/core';
import {
  Controller,
  CustomMethod,
  Delete,
  Get,
  Head,
  Next,
  nextFn,
  Patch,
  Post,
  Put,
  RestController,
} from '@gabliam/web-core';
import { WebPluginTest } from '@gabliam/web-core/lib/testing';
import ExpressPlugin, { express as e } from '../../src';
import { CUSTOM_ROUTER_CREATOR } from '../../src/constants';

let appTest: WebPluginTest;

beforeEach(async () => {
  appTest = new WebPluginTest([ExpressPlugin]);
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Integration Tests:', () => {
  [
    { decorator: Controller, name: 'Controller' },
    {
      decorator: RestController,
      name: 'RestController',
    },
  ].forEach(({ decorator, name }) => {
    describe(`decorator ${name}`, () => {
      test('should work with config', async () => {
        @decorator('rest.test.base')
        class TestController {
          @Get('rest.test.get')
          async getTest() {
            return 'config ok !!!!';
          }
        }

        appTest.addClass(TestController);
        appTest.addConf('rest.test', {
          base: '/test',
          get: '/',
        });
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/test')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest() {
            return new Promise(resolve => {
              setTimeout(resolve, 100, 'GET');
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async controller methods that fails', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest() {
            return new Promise((_, reject) => {
              setTimeout(reject, 100, 'GET');
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(500);
        expect(response).toMatchSnapshot();
      });

      test('should work for methods which call nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(@Next() nextFunc: nextFn) {
            return nextFunc();
          }

          @Get('/')
          public getTest2() {
            return 'GET';
          }
        }

        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async methods which call nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(@Next() nextFunc: nextFn) {
            return new Promise(resolve => {
              setTimeout(() => {
                resolve(nextFunc());
              }, 100);
            });
          }

          @Get('/')
          public getTest2() {
            return 'GET';
          }
        }
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for async methods called by nextFunc()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest(@Next() nextFunc: nextFn) {
            return nextFunc();
          }

          @Get('/')
          public getTest2() {
            return new Promise(resolve => {
              setTimeout(() => resolve('GET'), 100);
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should work for each shortcut decorator', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest() {
            return 'GET';
          }
          @Post('/')
          public postTest() {
            return 'POST';
          }
          @Put('/')
          public putTest() {
            return 'PUT';
          }
          @Patch('/')
          public patchTest() {
            return 'PATCH';
          }
          @Head('/')
          public headTest() {
            return 'HEAD';
          }
          @Delete('/')
          public deleteTest() {
            return 'DELETE';
          }
        }
        appTest.addClass(TestController);
        await appTest.buildAndStart();
        const agent = appTest.supertest();

        const rd = await agent.delete('/').expect(200);
        expect(rd).toMatchSnapshot();

        const rh = await agent.head('/').expect(200);
        expect(rh).toMatchSnapshot();

        const rpatch = await agent.patch('/').expect(200);
        expect(rpatch).toMatchSnapshot();

        const rput = await agent.put('/').expect(200);
        expect(rput).toMatchSnapshot();

        const rpost = await agent.post('/').expect(200);
        expect(rpost).toMatchSnapshot();

        const rget = await agent.get('/').expect(200);
        expect(rget).toMatchSnapshot();
      });

      test('should work for more obscure HTTP methods using the httpMethod decorator', async () => {
        @decorator('/')
        class TestController {
          @CustomMethod('propfind', '/')
          public getTest() {
            return 'PROPFIND';
          }
        }

        appTest.addClass(TestController);
        await appTest.buildAndStart();

        const response = await appTest
          .supertest()
          .propfind('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should use returned values as response', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          public getTest() {
            return { hello: 'world' };
          }
        }

        appTest.addClass(TestController);
        await appTest.buildAndStart();

        const response = await appTest
          .supertest()
          .get('/')
          .expect(200);
        expect(response).toMatchSnapshot();
      });

      test('should use custom router passed from configuration', async () => {
        @decorator('/CaseSensitive')
        class TestController {
          @Get('/Endpoint')
          public get() {
            return 'Such Text';
          }
        }

        @Config()
        class Conf {
          @Bean(CUSTOM_ROUTER_CREATOR)
          custom() {
            return () =>
              e.Router({
                caseSensitive: true,
              });
          }
        }
        appTest.addClass(Conf);
        appTest.addClass(TestController);
        await appTest.buildAndStart();

        const agent = appTest.supertest();

        const expectedSuccess = await agent
          .get('/CaseSensitive/Endpoint')
          .expect(200);

        expect(expectedSuccess).toMatchSnapshot();

        const expectedNotFound1 = await agent
          .get('/casesensitive/endpoint')
          .expect(404);

        expect(expectedNotFound1).toMatchSnapshot();

        const expectedNotFound2 = await agent
          .get('/CaseSensitive/endpoint')
          .expect(404);

        expect(expectedNotFound2).toMatchSnapshot();
      });

      test('should use custom routing configuration', async () => {
        @decorator('/ping')
        class TestController {
          @Get('/endpoint')
          public get() {
            return 'pong';
          }
        }

        appTest.addClass(TestController);
        appTest.addConf('application.web.rootPath', '/api/v1');
        appTest.addConf('application.web.bodyParser.text', true);
        await appTest.buildAndStart();

        const response = await appTest
          .supertest()
          .get('/api/v1/ping/endpoint')
          .expect(200);

        expect(response).toMatchSnapshot();
      });
    });
  });
});
