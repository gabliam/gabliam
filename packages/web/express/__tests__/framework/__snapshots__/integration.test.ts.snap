// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "Such Text",
}
`;

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /casesensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /CaseSensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "pong",
}
`;

exports[`Integration Tests: decorator Controller should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator Controller should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "130",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 500,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>GET</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "PROPFIND",
}
`;

exports[`Integration Tests: decorator Controller should work with config 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "14",
    "content-type": "text/html; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "config ok !!!!",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "11",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"Such Text\\"",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /casesensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /CaseSensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator RestController should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"pong\\"",
}
`;

exports[`Integration Tests: decorator RestController should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator RestController should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "28",
    "content-security-policy": "default-src 'none'",
    "content-type": "application/json; charset=utf-8",
    "x-content-type-options": "nosniff",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 500,
  "text": "{\\"status\\":500,\\"error\\":\\"GET\\"}",
}
`;

exports[`Integration Tests: decorator RestController should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "\\"DELETE\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "7",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "\\"PATCH\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PUT",
  },
  "status": 200,
  "text": "\\"PUT\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "POST",
  },
  "status": 200,
  "text": "\\"POST\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "10",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "\\"PROPFIND\\"",
}
`;

exports[`Integration Tests: decorator RestController should work with config 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "16",
    "content-type": "application/json; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "\\"config ok !!!!\\"",
}
`;
